File: sftp_transfer.pl
#!/usr/bin/env perl

=head1 NAME

sftp_transfer.pl - Transfer files via SFTP with verification and cleanup

=head1 SYNOPSIS

    sftp_transfer.pl [options] <local_file> <remote_file>

    Options:
        -H, --host <hostname>       Remote server hostname or IP (required)
        -u, --username <username>   Username for authentication (required)
        -p, --password <password>   Password for authentication
        -k, --key-file <path>       Path to SSH private key file
        -P, --port <port>           SSH/SFTP port (default: 22)
        --no-remove                 Do not remove source file after transfer
        -c, --checksum <algorithm>  Enable checksum verification (md5, sha1, sha256)
        -v, --verbose               Enable verbose logging
        -h, --help                  Show this help message

=head1 DESCRIPTION

This script transfers a file to a remote server using SFTP, verifies the
transfer was successful by comparing file sizes and optionally checksums,
and removes the source file after successful verification.

=head1 EXAMPLES

    # Transfer using password authentication
    sftp_transfer.pl -H server.example.com -u username -p password /path/to/file.txt /remote/path/file.txt

    # Transfer using SSH key
    sftp_transfer.pl -H server.example.com -u username -k ~/.ssh/id_rsa /path/to/file.txt /remote/path/file.txt

    # Transfer to a different port
    sftp_transfer.pl -H server.example.com -P 2222 -u username -p password /path/to/file.txt /remote/path/file.txt

    # Transfer with SHA256 checksum verification
    sftp_transfer.pl -H server.example.com -u username -p password -c sha256 /path/to/file.txt /remote/path/file.txt

=cut

use strict;
use warnings;
use Getopt::Long qw(:config no_ignore_case);
use Pod::Usage;
use File::Basename;
use File::Spec;
use Digest::MD5;
use Digest::SHA;

# Check for required modules
BEGIN {
    eval {
        require Net::SFTP::Foreign;
        Net::SFTP::Foreign->import();
    };
    if ($@) {
        print STDERR "Error: Net::SFTP::Foreign module is required.\n";
        print STDERR "Install it with: cpan Net::SFTP::Foreign\n";
        print STDERR "Or on Debian/Ubuntu: apt-get install libnet-sftp-foreign-perl\n";
        exit 1;
    }
}

# Global variables
my $verbose = 0;

# Parse command line options
my %opts = (
    port => 22,
    help => 0,
    'no-remove' => 0,
    verbose => 0,
    checksum => undef,
);

GetOptions(
    'H|host=s'      => \$opts{host},
    'u|username=s'  => \$opts{username},
    'p|password=s'  => \$opts{password},
    'k|key-file=s'  => \$opts{key_file},
    'P|port=i'      => \$opts{port},
    'no-remove'     => \$opts{'no-remove'},
    'c|checksum=s'  => \$opts{checksum},
    'v|verbose'     => \$opts{verbose},
    'h|help'        => \$opts{help},
) or pod2usage(2);

# Show help if requested
pod2usage(-verbose => 2, -exitval => 0) if $opts{help};

# Set verbose flag
$verbose = $opts{verbose};

# Validate required arguments
my $local_file = shift @ARGV;
my $remote_file = shift @ARGV;

unless (defined $local_file && defined $remote_file) {
    print STDERR "Error: Both local_file and remote_file arguments are required\n\n";
    pod2usage(2);
}

unless (defined $opts{host}) {
    print STDERR "Error: --host is required\n\n";
    pod2usage(2);
}

unless (defined $opts{username}) {
    print STDERR "Error: --username is required\n\n";
    pod2usage(2);
}

unless (defined $opts{password} || defined $opts{key_file}) {
    log_error("Either --password or --key-file must be provided");
    exit 1;
}

# Validate checksum algorithm if provided
if (defined $opts{checksum}) {
    unless ($opts{checksum} =~ /^(md5|sha1|sha256)$/i) {
        log_error("Invalid checksum algorithm: $opts{checksum}. Must be one of: md5, sha1, sha256");
        exit 1;
    }
    $opts{checksum} = lc($opts{checksum});
    log_info("Checksum verification enabled: $opts{checksum}");
}

# Validate local file exists
unless (-f $local_file) {
    log_error("Local file not found: $local_file");
    exit 1;
}

# Main execution
exit main();

sub main {
    eval {
        # Step 1: Connect to remote server
        log_info("Connecting to $opts{host}:$opts{port} as $opts{username}");
        my $sftp = connect_sftp();
        unless ($sftp) {
            log_error("Failed to establish SFTP connection");
            return 1;
        }
        
        # Step 2: Transfer the file
        log_info("Transferring $local_file to $remote_file");
        unless (transfer_file($sftp, $local_file, $remote_file)) {
            log_error("File transfer failed");
            return 1;
        }
        
        # Step 3: Verify the transfer
        log_info("Verifying file transfer...");
        unless (verify_transfer($sftp, $local_file, $remote_file)) {
            log_error("File transfer verification failed");
            return 1;
        }
        
        # Step 4: Remove source file (unless --no-remove flag is set)
        unless ($opts{'no-remove'}) {
            unless (remove_source_file($local_file)) {
                log_error("Failed to remove source file");
                return 1;
            }
        } else {
            log_info("Skipping source file removal (--no-remove flag set)");
        }
        
        log_success("All operations completed successfully");
        return 0;
    };
    
    if ($@) {
        log_error("Unexpected error: $@");
        return 1;
    }
}

sub connect_sftp {
    my %sftp_opts = (
        host => $opts{host},
        user => $opts{username},
        port => $opts{port},
    );
    
    # Add authentication method
    if (defined $opts{key_file}) {
        log_info("Using key file: $opts{key_file}");
        $sftp_opts{key_path} = $opts{key_file};
    } elsif (defined $opts{password}) {
        log_info("Using password authentication");
        $sftp_opts{password} = $opts{password};
    }
    
    # Additional options for better compatibility
    # WARNING: StrictHostKeyChecking=no disables host key verification
    # This makes the connection vulnerable to man-in-the-middle attacks
    # In production, you should properly manage known_hosts or use custom verification
    $sftp_opts{more} = ['-o', 'StrictHostKeyChecking=no'];
    
    my $sftp = Net::SFTP::Foreign->new(%sftp_opts);
    
    if ($sftp->error) {
        log_error("Connection failed: " . $sftp->error);
        return undef;
    }
    
    log_info("SFTP connection established");
    return $sftp;
}

sub transfer_file {
    my ($sftp, $local, $remote) = @_;
    
    # Get local file size
    my $local_size = -s $local;
    log_info("File size: $local_size bytes");
    
    # Create remote directory if needed
    my $remote_dir = dirname($remote);
    if ($remote_dir && $remote_dir ne '' && $remote_dir ne '.' && $remote_dir ne '/') {
        log_debug("Ensuring remote directory exists: $remote_dir");
        $sftp->mkpath($remote_dir);
        if ($sftp->error) {
            log_error("Failed to create remote directory: " . $sftp->error);
            return 0;
        }
    }
    
    # Transfer the file
    $sftp->put($local, $remote);
    
    if ($sftp->error) {
        log_error("Transfer failed: " . $sftp->error);
        return 0;
    }
    
    log_info("File transfer completed");
    return 1;
}

sub calculate_local_checksum {
    my ($file, $algorithm) = @_;
    
    log_debug("Calculating $algorithm checksum for local file: $file");
    
    open(my $fh, '<', $file) or do {
        log_error("Cannot open file for checksum: $!");
        return undef;
    };
    binmode($fh);
    
    my $digest;
    if ($algorithm eq 'md5') {
        $digest = Digest::MD5->new;
    } elsif ($algorithm eq 'sha1') {
        $digest = Digest::SHA->new(1);
    } elsif ($algorithm eq 'sha256') {
        $digest = Digest::SHA->new(256);
    } else {
        log_error("Unsupported checksum algorithm: $algorithm");
        close($fh);
        return undef;
    }
    
    $digest->addfile($fh);
    close($fh);
    
    my $checksum = $digest->hexdigest;
    log_debug("Local $algorithm checksum: $checksum");
    return $checksum;
}

sub calculate_remote_checksum {
    my ($sftp, $remote_file, $algorithm) = @_;
    
    log_debug("Calculating $algorithm checksum for remote file: $remote_file");
    
    # Map algorithm to the appropriate command
    my $cmd;
    if ($algorithm eq 'md5') {
        # Try different MD5 commands (md5sum on Linux, md5 on macOS/BSD)
        $cmd = "(md5sum '$remote_file' 2>/dev/null || md5 -r '$remote_file' 2>/dev/null) | awk '{print \$1}'";
    } elsif ($algorithm eq 'sha1') {
        $cmd = "(sha1sum '$remote_file' 2>/dev/null || shasum -a 1 '$remote_file' 2>/dev/null) | awk '{print \$1}'";
    } elsif ($algorithm eq 'sha256') {
        $cmd = "(sha256sum '$remote_file' 2>/dev/null || shasum -a 256 '$remote_file' 2>/dev/null) | awk '{print \$1}'";
    } else {
        log_error("Unsupported checksum algorithm: $algorithm");
        return undef;
    }
    
    # Execute command via SSH
    my $output = $sftp->system($cmd);
    
    if ($sftp->error) {
        log_error("Failed to calculate remote checksum: " . $sftp->error);
        return undef;
    }
    
    # Read the output
    my $checksum;
    if (defined $output) {
        $checksum = $output;
        $checksum =~ s/^\s+|\s+$//g;  # Trim whitespace
    }
    
    # The system() method doesn't return command output directly in Net::SFTP::Foreign
    # We need to use a different approach - download a small chunk or use a temp file
    # Let's use a more reliable method with backtick-style capture
    
    # Create a temporary file to capture output
    my $temp_remote = "/tmp/.sftp_checksum_$$";
    my $temp_local = "/tmp/.sftp_checksum_local_$$";
    
    # Execute command and redirect output to temp file
    $sftp->system("$cmd > $temp_remote 2>&1");
    
    if ($sftp->error) {
        log_error("Failed to execute remote checksum command: " . $sftp->error);
        return undef;
    }
    
    # Download the temp file
    $sftp->get($temp_remote, $temp_local);
    
    if ($sftp->error) {
        log_error("Failed to retrieve checksum result: " . $sftp->error);
        # Clean up remote temp file
        $sftp->system("rm -f $temp_remote");
        return undef;
    }
    
    # Read the checksum from local temp file
    if (open(my $fh, '<', $temp_local)) {
        $checksum = <$fh>;
        close($fh);
        if (defined $checksum) {
            $checksum =~ s/^\s+|\s+$//g;  # Trim whitespace
        }
        unlink($temp_local);
    } else {
        log_error("Failed to read checksum result: $!");
        $sftp->system("rm -f $temp_remote");
        return undef;
    }
    
    # Clean up remote temp file
    $sftp->system("rm -f $temp_remote");
    
    unless ($checksum && $checksum =~ /^[a-f0-9]+$/i) {
        log_error("Invalid checksum format received: " . ($checksum || 'empty'));
        return undef;
    }
    
    log_debug("Remote $algorithm checksum: $checksum");
    return lc($checksum);
}

sub verify_transfer {
    my ($sftp, $local, $remote) = @_;
    
    # Get remote file attributes
    my $remote_attrs = $sftp->stat($remote);
    
    unless ($remote_attrs) {
        log_error("Remote file not found: $remote");
        log_error("SFTP error: " . $sftp->error) if $sftp->error;
        return 0;
    }
    
    # Get file sizes
    my $local_size = -s $local;
    my $remote_size = $remote_attrs->{size};
    
    log_info("Local file size: $local_size bytes");
    log_info("Remote file size: $remote_size bytes");
    
    # Compare sizes
    if ($local_size != $remote_size) {
        log_error("File sizes do not match!");
        return 0;
    }
    
    log_success("File size verification passed");
    
    # Perform checksum verification if requested
    if (defined $opts{checksum}) {
        log_info("Performing checksum verification using $opts{checksum}...");
        
        my $local_checksum = calculate_local_checksum($local, $opts{checksum});
        unless (defined $local_checksum) {
            log_error("Failed to calculate local checksum");
            return 0;
        }
        
        my $remote_checksum = calculate_remote_checksum($sftp, $remote, $opts{checksum});
        unless (defined $remote_checksum) {
            log_error("Failed to calculate remote checksum");
            return 0;
        }
        
        log_info("Local checksum:  $local_checksum");
        log_info("Remote checksum: $remote_checksum");
        
        if ($local_checksum ne $remote_checksum) {
            log_error("Checksum verification failed! Files do not match.");
            return 0;
        }
        
        log_success("Checksum verification passed");
    }
    
    log_success("File transfer verified successfully");
    return 1;
}

sub remove_source_file {
    my ($file) = @_;
    
    log_info("Removing source file: $file");
    
    if (unlink($file)) {
        log_success("Source file removed successfully");
        return 1;
    } else {
        log_error("Failed to remove source file: $!");
        return 0;
    }
}

# Logging functions
sub log_info {
    my ($message) = @_;
    my $timestamp = get_timestamp();
    print "$timestamp - INFO - $message\n";
}

sub log_error {
    my ($message) = @_;
    my $timestamp = get_timestamp();
    print STDERR "$timestamp - ERROR - $message\n";
}

sub log_success {
    my ($message) = @_;
    my $timestamp = get_timestamp();
    print "$timestamp - SUCCESS - ✓ $message\n";
}

sub log_debug {
    my ($message) = @_;
    return unless $verbose;
    my $timestamp = get_timestamp();
    print "$timestamp - DEBUG - $message\n";
}

sub get_timestamp {
    my ($sec, $min, $hour, $mday, $mon, $year) = localtime(time);
    return sprintf("%04d-%02d-%02d %02d:%02d:%02d",
                   $year + 1900, $mon + 1, $mday, $hour, $min, $sec);
}

__END__

=head1 EXIT CODES

=over 4

=item * 0 - Success: file transferred, verified, and removed

=item * 1 - Failure: see error messages for details

=back

=head1 AUTHOR

Created for the ssh-transfer project

=head1 LICENSE

MIT License

=cut

File: sftp_transfer.py
#!/usr/bin/env python3
"""
SFTP File Transfer Script

This script transfers a file to a remote server using SFTP,
verifies the transfer was successful, and removes the source file.
"""

import argparse
import os
import sys
import logging
from pathlib import Path

try:
    import paramiko
except ImportError:
    print("Error: paramiko library is required. Install it with: pip install paramiko")
    sys.exit(1)


# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class SFTPTransfer:
    """Handle SFTP file transfer operations."""
    
    def __init__(self, hostname, port, username, password=None, key_file=None):
        """
        Initialize SFTP connection parameters.
        
        Args:
            hostname: Remote server hostname or IP
            port: SSH/SFTP port (default 22)
            username: Username for authentication
            password: Password for authentication (optional)
            key_file: Path to private key file for authentication (optional)
        """
        self.hostname = hostname
        self.port = port
        self.username = username
        self.password = password
        self.key_file = key_file
        self.client = None
        self.sftp = None
    
    def connect(self):
        """Establish SFTP connection to remote server."""
        try:
            logger.info(f"Connecting to {self.hostname}:{self.port} as {self.username}")
            
            self.client = paramiko.SSHClient()
            self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            
            # Prepare connection parameters
            connect_kwargs = {
                'hostname': self.hostname,
                'port': self.port,
                'username': self.username,
            }
            
            # Add authentication method
            if self.key_file:
                logger.info(f"Using key file: {self.key_file}")
                connect_kwargs['key_filename'] = self.key_file
            elif self.password:
                logger.info("Using password authentication")
                connect_kwargs['password'] = self.password
            else:
                logger.error("No authentication method provided (password or key file)")
                return False
            
            self.client.connect(**connect_kwargs)
            self.sftp = self.client.open_sftp()
            logger.info("SFTP connection established")
            return True
            
        except Exception as e:
            logger.error(f"Failed to connect: {e}")
            return False
    
    def transfer_file(self, local_path, remote_path):
        """
        Transfer a file to the remote server.
        
        Args:
            local_path: Path to local file
            remote_path: Destination path on remote server
            
        Returns:
            bool: True if transfer successful, False otherwise
        """
        try:
            if not os.path.exists(local_path):
                logger.error(f"Local file not found: {local_path}")
                return False
            
            local_size = os.path.getsize(local_path)
            logger.info(f"Transferring {local_path} ({local_size} bytes) to {remote_path}")
            
            # Create remote directory if it doesn't exist
            remote_dir = os.path.dirname(remote_path)
            if remote_dir:
                try:
                    self.sftp.stat(remote_dir)
                except FileNotFoundError:
                    logger.info(f"Creating remote directory: {remote_dir}")
                    self._mkdir_recursive(remote_dir)
            
            # Transfer the file
            self.sftp.put(local_path, remote_path)
            logger.info("File transfer completed")
            return True
            
        except Exception as e:
            logger.error(f"File transfer failed: {e}")
            return False
    
    def _mkdir_recursive(self, remote_path):
        """Create remote directory recursively."""
        parts = remote_path.split('/')
        current_path = ''
        
        for part in parts:
            if not part:
                current_path = '/'
                continue
            
            current_path = os.path.join(current_path, part)
            try:
                self.sftp.stat(current_path)
            except FileNotFoundError:
                self.sftp.mkdir(current_path)
    
    def verify_transfer(self, local_path, remote_path):
        """
        Verify that the file was transferred successfully.
        
        Checks:
        1. Remote file exists
        2. File sizes match
        
        Args:
            local_path: Path to local file
            remote_path: Path to remote file
            
        Returns:
            bool: True if verification successful, False otherwise
        """
        try:
            logger.info("Verifying file transfer...")
            
            # Check if remote file exists
            try:
                remote_stat = self.sftp.stat(remote_path)
            except FileNotFoundError:
                logger.error(f"Remote file not found: {remote_path}")
                return False
            
            # Compare file sizes
            local_size = os.path.getsize(local_path)
            remote_size = remote_stat.st_size
            
            logger.info(f"Local file size: {local_size} bytes")
            logger.info(f"Remote file size: {remote_size} bytes")
            
            if local_size != remote_size:
                logger.error("File sizes do not match!")
                return False
            
            logger.info("✓ File transfer verified successfully")
            return True
            
        except Exception as e:
            logger.error(f"Verification failed: {e}")
            return False
    
    def close(self):
        """Close SFTP and SSH connections."""
        if self.sftp:
            self.sftp.close()
        if self.client:
            self.client.close()
        logger.info("Connection closed")


def remove_source_file(file_path):
    """
    Remove the source file after successful transfer.
    
    Args:
        file_path: Path to file to remove
        
    Returns:
        bool: True if removal successful, False otherwise
    """
    try:
        logger.info(f"Removing source file: {file_path}")
        os.remove(file_path)
        logger.info("✓ Source file removed successfully")
        return True
    except Exception as e:
        logger.error(f"Failed to remove source file: {e}")
        return False


def main():
    """Main entry point for the script."""
    parser = argparse.ArgumentParser(
        description='Transfer a file using SFTP, verify transfer, and remove source file',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Transfer using password authentication
  %(prog)s -H server.example.com -u username -p password /path/to/file.txt /remote/path/file.txt
  
  # Transfer using SSH key
  %(prog)s -H server.example.com -u username -k ~/.ssh/id_rsa /path/to/file.txt /remote/path/file.txt
  
  # Transfer to a different port
  %(prog)s -H server.example.com -P 2222 -u username -p password /path/to/file.txt /remote/path/file.txt
        """
    )
    
    parser.add_argument('local_file', help='Path to the local file to transfer')
    parser.add_argument('remote_file', help='Destination path on the remote server')
    parser.add_argument('-H', '--host', required=True, help='Remote server hostname or IP')
    parser.add_argument('-P', '--port', type=int, default=22, help='SSH/SFTP port (default: 22)')
    parser.add_argument('-u', '--username', required=True, help='Username for authentication')
    parser.add_argument('-p', '--password', help='Password for authentication')
    parser.add_argument('-k', '--key-file', help='Path to SSH private key file')
    parser.add_argument('--no-remove', action='store_true', 
                       help='Do not remove source file after transfer (for testing)')
    parser.add_argument('-v', '--verbose', action='store_true', 
                       help='Enable verbose logging')
    
    args = parser.parse_args()
    
    # Set logging level
    if args.verbose:
        logger.setLevel(logging.DEBUG)
    
    # Validate authentication method
    if not args.password and not args.key_file:
        logger.error("Either --password or --key-file must be provided")
        return 1
    
    # Validate local file exists
    if not os.path.exists(args.local_file):
        logger.error(f"Local file not found: {args.local_file}")
        return 1
    
    # Initialize SFTP transfer
    sftp_transfer = SFTPTransfer(
        hostname=args.host,
        port=args.port,
        username=args.username,
        password=args.password,
        key_file=args.key_file
    )
    
    try:
        # Step 1: Connect to remote server
        if not sftp_transfer.connect():
            logger.error("Failed to establish SFTP connection")
            return 1
        
        # Step 2: Transfer the file
        if not sftp_transfer.transfer_file(args.local_file, args.remote_file):
            logger.error("File transfer failed")
            return 1
        
        # Step 3: Verify the transfer
        if not sftp_transfer.verify_transfer(args.local_file, args.remote_file):
            logger.error("File transfer verification failed")
            return 1
        
        # Step 4: Remove source file (unless --no-remove flag is set)
        if not args.no_remove:
            if not remove_source_file(args.local_file):
                logger.error("Failed to remove source file")
                return 1
        else:
            logger.info("Skipping source file removal (--no-remove flag set)")
        
        logger.info("✓ All operations completed successfully")
        return 0
        
    except KeyboardInterrupt:
        logger.warning("\nOperation cancelled by user")
        return 1
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        return 1
    finally:
        sftp_transfer.close()


if __name__ == '__main__':
    sys.exit(main())
